File: Data Structures & Algorithms/reorder-linked-list/submission-8.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> ListNode:
        prev, curr = None, head
        while curr:
            currNext = curr.next
            curr.next = prev
            prev = curr
            curr = currNext
        return prev

    def reorderList(self, head: Optional[ListNode]) -> None:
        slow, fast = head, head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        first = head
        second = slow.next
        slow.next = None
        second = self.reverseList(second)
        while first and second:
            t1, t2 = first.next, second.next
            first.next, second.next = second, t1
            first, second = t1, t2
        

File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-8.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    unordered_map<Node*, Node*> node_map;
    Node* copyRandomList(Node* head) {
        Node* curr = head;
        while (curr) {
            Node* new_node = new Node(curr->val);
            node_map[curr] = new_node;
            curr = curr->next;
        }
        for (Node* old_node = head; old_node; old_node = old_node->next) {
            node_map[old_node]->next = node_map[old_node->next];
            node_map[old_node]->random = node_map[old_node->random];
        }
        return node_map[head];
    }
};
